### PR TITLE
Detect and remove recursive calls

### LINKED ISSUE #1566: WASM-GEN: make infinite recursion not so often
### Problem to Solve

Currently `wasm-smith` generates too many codes, which has infinite recursion inside, this lead to the fact that in too many cases we finite our execution because of Stack Overflow error. 

### Possible Solution

We can restrict in most cases to call itself inside a function body:
```
(func 0
  ...
  call 0
  ...
)
```
I think this can significantly decrease number of smart contracts with such problem 

### Notes

_No response_

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1566 

- implement `find_recursion` method based on DFS
- implement and use `remove_recursion` procedure
- test for well-known base cases

@gear-tech/dev

### CODE REVIEW

File: utils/wasm-gen/Cargo.toml [hunk 1]
```diff
@@ -14,4 +14,5 @@ wasmprinter = "0.2"
 [dev-dependencies]
 rand = { version = "0.8.0", features = ["small_rng"] }
 wasmparser = { version = "0.93.0" }
+wat = "1.0.53"
```
Comment (shamilsan) [inline utils/wasm-gen/Cargo.toml:17]: The last version of `wat` is v1.0.56
```suggestion
wat = "1.0.56"
```

File: utils/wasm-gen/src/lib.rs [hunk 2]
```diff
@@ -135,6 +138,7 @@ impl GearConfig {
             skip_init: (1, 1000).into(),
             skip_handle: prob,
             skip_init_when_no_funcs: prob,
+            remove_recursion: (95, 100).into(),
```
Comment (shamilsan) [inline utils/wasm-gen/src/lib.rs:141]: Isn't is better to add a const?

File: utils/wasm-gen/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
```
Comment (shamilsan) [inline utils/wasm-gen/src/utils.rs:3]: Please turn the calendar over.
```suggestion
// Copyright (C) 2023 Gear Technologies Inc.
```

File: utils/wasm-gen/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    Callback: FnMut(&[usize], usize),
+{
+    let function_bodies = match module.code_section() {
+        Some(s) if !s.bodies().is_empty() => s.bodies(),
+        _ => return,
+    };
+
+    let import_count = module.import_count(ImportCountType::Function);
+
+    let mut colored_list = Vec::<HashMap<_, _>>::with_capacity(function_bodies.len());
```
Comment (shamilsan) [inline utils/wasm-gen/src/utils.rs:48]: Is this type annotation really needed?
```suggestion
    let mut colored_list = Vec::with_capacity(function_bodies.len());
```

**Review by shamilsan — APPROVED** [review]
Awesome 🔥

File: utils/wasm-gen/src/lib.rs (hunk 2 above)
Comment (gshep) [reply utils/wasm-gen/src/lib.rs:141]: It makes sense however I'm not sure about the case. Frankly speaking the value is also a subject to discuss

**Review by breathx — APPROVED** [review]
@grishasobol review please

File: utils/wasm-gen/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    }
+}
+
+fn find_recursion_impl<Callback>(
+    call_index: usize,
+    import_count: usize,
+    function_bodies: &[FuncBody],
+    colored: &mut HashMap<usize, Color>,
+    path: &mut Vec<usize>,
+    callback: &mut Callback,
```
[the remark this answers is not in the document]
Comment (gshep) [reply utils/wasm-gen/src/utils.rs:80]: it does not compiled without mut

File: utils/wasm-gen/src/lib.rs [hunk 6]
```diff
@@ -734,7 +740,7 @@ impl<'a> WasmGen<'a> {
             }
         }
 
-        module
+        utils::remove_recursion(module)
```
[the remark this answers is not in the document]
Comment (gshep) [reply utils/wasm-gen/src/lib.rs:743]: good catch, thanks!

File: utils/wasm-gen/src/test.rs [hunk 7]
```diff
@@ -64,3 +65,48 @@ fn gen_wasm_valid() {
...
+        )
+        (func (;2;)
+            call 1
+            drop
+        )
+    )"#;
+
+    let wasm = wat::parse_str(wat).unwrap();
+    let module: elements::Module = parity_wasm::deserialize_buffer(&wasm).unwrap();
+    utils::find_recursion(&module, |path, call| {
```
[the remark this answers is not in the document]
Comment (gshep) [reply utils/wasm-gen/src/test.rs:103]: ok

File: utils/wasm-gen/src/test.rs [hunk 8]
```diff
@@ -64,3 +65,52 @@ fn gen_wasm_valid() {
         wasmparser::validate(&code).unwrap();
     }
 }
+
+#[test]
```
Comment (grishasobol) [inline utils/wasm-gen/src/test.rs:69]: Can you move this to tests.rs  if it is possible?

File: utils/wasm-gen/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                    .with_instructions(elements::Instructions::new(body))
+                    .build()
+                    .build(),
+            )
+            .body;
+
+        call_substitutions.insert(call_index, mock_index + import_count as u32);
+    }
+
+    // change call indices to mock functions to disable recursion
```
Comment (grishasobol) [inline utils/wasm-gen/src/utils.rs:192]: indexes?

**Review by grishasobol — COMMENTED** [review]
You need to pass `gear-wasm-gen::tests::*` before approve

File: utils/wasm-gen/src/utils.rs (hunk 9 above)
Comment (gshep) [reply utils/wasm-gen/src/utils.rs:192]: if i'm not mistaken both are correct

**Review by SabaunT — APPROVED** [review]

**Pull request merged**